### PR TITLE
Fix mirostat state when using multiple sequences

### LINKED ISSUE #3537: [Bug] Mirostat samplers don't work properly with parallel generation
This is because `llama_sample_token` in `common.cpp` uses a static for mirostat1 and 2 `mu`. Because of this, different sequences will affect each other (including ones that were already deleted). 

The fix for this doesn't really seem that simple. I don't think it can be done only inside `llama_sample_token`. I think `llama_sample_token` is going to have to get changed to take something like a sequence-specific sampler state structure where stuff like that sequence's `mu` could get stored. Then it would be up to the app to reset `mu` when appropriate (like the sequence ends and the slot will be reused).

### PR DESCRIPTION
**Pull request by KerfuffleV2**
As mentioned in #3537, mirostat currently isn't compatible with using multiple sequences.

The main selling point of the way this pull is implemented is such a way that it's pretty simple and uninvasive.

However, I _really_ don't like storing mutable sampler state in `gpt_params` (even though it's only in common and not the main `llama.cpp` API). This also required removing the `const` from the `params` argument to `llama_sample_token`. As far as I can see, the existing examples don't care about that.

I feel like the right way to do this is probably to move the sampler state out of `gpt_params` and have it passed separately. In that case, this is probably also where grammar should be since it's a type of sampler state. So we wouldn't add a new argument to `llama_sample_token`, we'd replace the current grammar one with sampler state. This of course would require changing a lot more stuff, including all the examples that use `llama_sample_token` (I don't think it would be _too_ bad though).

Thoughts?

Closes #3537

### CODE REVIEW

**Review by ggerganov — APPROVED** [review]
Yes, dedicated sampling state with grammar and mirostat would be better. Maybe implemented in `common/sampling.h/.cpp`. It should probably inherit all sampling-related parameters `gpt_params`, such as temperature, top_p, top_k, etc, so that `llama_sample_token` accepts `struct llama_sampling_state` instead of `struct gpt_params`.

For now we can have this workaround

**Comment by KerfuffleV2** [comment]
@ggerganov 

> For now we can have this workaround

Do you actually prefer doing it this way for now? 

I don't mind changing this to do it the other way I suggested as long as you agree that approach is okay.

> Maybe implemented in common/sampling.h/.cpp. 

This actually raises another question that I'm actually dealing with in my seqrep sampler. Right now it's really awkward to have multiple source files in common. This is how I dealt with it:

```makefile
COMMON_DEPS = common/common.cpp common/common.h build-info.h common/log.h
COMMON_OBJS = common.o
ifndef LLAMA_DISABLE_SEQREP_SAMPLER
COMMON_DEPS += common/seqrep-sampler.cpp common/seqrep-sampler.h
COMMON_OBJS += seqrep-sampler.o
endif
common.o: $(COMMON_DEPS)
	$(CXX) $(CXXFLAGS) -c $< -o $@

simple: examples/simple/simple.cpp                            build-info.h ggml.o llama.o $(COMMON_OBJS) $(OBJS)
	$(CXX) $(CXXFLAGS) $(filter-out %.h,$^) -o $@ $(LDFLAGS)
```

and so on. It's probably less of a pain in cmake.

**Comment by ggerganov** [comment]
We should try separating the sampling state - would be better than the current fix, so let's give it a try if you are up to.

The proposed Makefile looks OK to me.

**Comment by FSSRepo** [comment]
When this commit is merged into the master branch, we need to add the parameter `llama_sample_token(ctx, NULL, NULL, params, slot.tokens_prev, candidates, slot.i_batch, slot.seq_id)` in the `llama_sample_token` call of the `server-parallel` in #3490.

**Comment by KerfuffleV2** [comment]
Well, this went from small and self-contained to huge and complicated. I sure hope I'm on the right track after all this.

Pretty much all the sampling stuff got moved into `common/sampling.{c,h}`. Creating sampling state takes `gpt_params` and `llama_grammar *` (can be NULL). The sampling-related params in `gpt_params` are also in a `llama_sampling_params` struct now.

The sampling state (`llama_sampling_state`) holds a copy of the params that were created at init time. It also holds per-sequence state for samplers that use it (Mirostat 1/2, and grammar).

If the per-sequence state doesn't exist when `llama_sample_token` is called, it will be created with default values. In the case of grammar, this does `llama_grammar_copy` on the `llama_grammar *` that was supplied at init time.

If you want to use separate grammar, or separate grammar states per sequence then you'd probably have to manually manage the grammar part yourself. (Not sure this is so easy right now, so I might need to add an interface.)

In the case of parallel generation, when a sequence ends and you will want to reuse it (or just free up memory) you need to call `llama_sampling_state_reset` with the sequence id. This will reset stuff like Mirostat `mu` to the initial value so it won't mess with future generation with that sequence id.

I also randomly threw in support for `LLAMA_SANITIZE_{THREAD,ADDRESS,UNDEFINED}` in the `Makefile`.

I think this currently doesn't break stuff, but there were some tricky parts like `server` and `speculative`.

**Comment by KerfuffleV2** [comment]
@FSSRepo You're going to hate me when you see the next step in this pull.

Calling sampling is going to look like:

```c++
llama_sample_token(ctx, NULL, sampling_state, slot.tokens_prev, candidates, slot.i_batch, slot.id);
```

It looks like #3490 doesn't support grammar currently? So that's going to make your life easier. Pretty much the only other thing to worry about is calling `llama_sampling_state_reset`  when you're done with a sequence id (but may want to reuse it for a different generation). Pretty much when you hit the EOS token or reach the quota of tokens to generate, you can just reset that sequence id.

**Comment by KerfuffleV2** [comment]
I exported the function to fetch/create default instances of sampler state. This should fix the problem I mentioned earlier about how it would be hard to do something like parallel generation where each sequence used its own grammar.

By the way, since the `ggml-alloc` stuff:

```plaintext
ggml-alloc.c:212:32: runtime error: pointer index expression with base 0x00000100a020 overflowed to 0xffffffffffffffff
llama_new_context_with_model: compute buffer total size = 552.88 MB
llama_new_context_with_model: VRAM scratch buffer: 546.75 MB
llama_new_context_with_model: total VRAM used: 6515.44 MB (model: 3368.69 MB, context: 3146.75 MB)
ggml-cuda.cu:6787:51: runtime error: applying non-zero offset 1152 to null pointer
```

Not sure if that's anything to worry about. The `ggml-alloc` one is new, the other one isn't. (I was assuming it was just because GCC's address sanitizing stuff doesn't know about CUDA/ROCM.)

**Comment by KerfuffleV2** [comment]
This isn't really approved, right? Even without a full review, any changes I can/should start working on?

File: common/sampling.cpp [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+    state.sequence_states.erase(it);
+    return true;
+}
+
+llama_token llama_sample_token(
+                  struct llama_context * ctx,
+                  struct llama_context * ctx_guidance,
+                  struct llama_sampling_state & state,
+        const std::vector<llama_token> & last_tokens,
+         std::vector<llama_token_data> & candidates,
```
Comment (ggerganov) [inline common/sampling.cpp:55]: Should we absorb `last_tokens` and `candidates` into `llama_sampling_state`?

File: common/sampling.cpp [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+            id = llama_sample_token(ctx, &cur_p);
+
+            LOG("sampled token: %5d: '%s'\n", id, llama_token_to_piece(ctx, id).c_str());
+        }
+    }
+
+    if (seq_state.grammar != NULL) {
+        llama_grammar_accept_token(ctx, seq_state.grammar, id);
+    }
```
Comment (ggerganov) [inline common/sampling.cpp:163]: Probably we should add `llama_sampling_accept_token()` and move this call in there together with update of `last_tokens` member (if we decide it should become part of `llama_sampling_state`)

**Review by ggerganov — APPROVED** [review]
This is a great change. I actually think that we should merge `llama_sampling` directly into `llama.cpp`. But let's do this after this PR is merged and tested for some time

> Not sure if that's anything to worry about.

These errors look benign, but we will look in ways to fix them anyway

File: common/sampling.h [hunk 3]
```diff
@@ -0,0 +1,110 @@
...
+//  - grammar:       grammar to use for sampling, ignore if NULL
+//  - last_tokens:   needed for repetition penalty, ignore if empty
+//  - idx:           sample from llama_get_logits_ith(ctx, idx)
+//  - seq:           sequence id to associate sampler state with
+//
+// returns:
+//  - token:      sampled token
+//  - candidates: vector of candidate tokens
+//
+llama_token llama_sample_token(
```
Comment (ggerganov) [inline common/sampling.h:103]:
```suggestion
llama_token llama_sampling_sample(
```

File: common/sampling.h [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+} llama_sampling_params;
+
+// per-sequence sampler state
+typedef struct llama_sampler_sequence_state {
+    float mirostat_mu; // mirostat sampler state
+    llama_grammar * grammar;
+} llama_sampler_sequence_state;
+
+// general sampler state
+typedef struct llama_sampling_state {
```
Comment (ggerganov) [inline common/sampling.h:44]: We are leaning towards using the term `context` when we create objects containing some sort of state. So for the sake of consistency in the codebase, we should probably rename this to `llama_sampling_context`

File: common/sampling.cpp (hunk 1 above)
Comment (KerfuffleV2) [reply common/sampling.cpp:55]: Last tokens can be specific to the sequence, right? So this would kind of mean stuff using last tokens would have to be aware of sequences. I kind of feel like this also might limit how people can manipulate last tokens and if there are currently examples that do that kind of thing it might be difficult to adapt them (for me anyway, since I'm not really deeply familiar with most of them). 

`candidates` I'm less sure about, it's basically just a scratch area for the logits in a form samplers can work with (right?) so I think moving it in there is less of a big deal. It's a pretty large structure though, I don't know if that's a consideration. Right now the current stuff in those structs is pretty lightweight.

I don't have very strong feelings about this. I'd like to say "These changes are already complicated enough, let's come back to that" but... I probably never would. :)

File: common/sampling.h (hunk 3 above)
Comment (KerfuffleV2) [reply common/sampling.h:103]: I really like this for reasons other than just consistency. I've found it confusing in the past that `llama_sample_token` was an overload in `common` while there was also a `llama_sample_token` in `lamma.cpp`.

File: common/sampling.h (hunk 4 above)
Comment (KerfuffleV2) [reply common/sampling.h:44]: Also `llama_sampler_sequence_context` I assume.

**Comment by KerfuffleV2** [comment]
Current status: I took the suggestions to rename the functions/types but I didn't do stuff like moving `last_tokens` into the sampling context (yet).

File: common/sampling.cpp [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+        it->second.grammar = NULL;
+    }
+    sampling_ctx.sequence_contexts.erase(it);
+    return true;
+}
+
+llama_token llama_sampling_sample(
+                  struct llama_context * ctx,
+                  struct llama_context * ctx_guidance,
+                  struct llama_sampling_context & sampling_ctx,
```
Comment (ggerganov) [inline common/sampling.cpp:53]:
```suggestion
                  struct llama_sampling_context & ctx_sampling,
```

There are a few other places that need similar change for consistency sake

Comment (KerfuffleV2) [reply common/sampling.cpp]: By "a few" you mean 20 or so? :) Hopefully I caught them all. Everything seems to compile/work still.

**Review by ggerganov — APPROVED** [review]

**Comment by ggerganov** [comment]
Thanks for this - I accidentally merged this too quickly with the old title. Should have updated to the more relevant change of introducing the `llama_sampling_context`.

File: common/sampling.h [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+typedef struct llama_sampling_context {
+    ~llama_sampling_context();
+
+    // parameters that will be used for sampling and when creating
+    // new llama_sampler_sequence_context instances
+    llama_sampling_params params;
+
+    // map of sequence ids to sampler contexts
+    std::unordered_map<llama_seq_id, llama_sampler_sequence_context> sequence_contexts;
+
```
Comment (ggerganov) [inline common/sampling.h:53]: Any reason not to have single sequence data in `llama_sampling_context`?
When we want to sample multiple sequences, we will create on `llama_sampling_context` for each.
